src/vdom/diff.js

```diff
@@ -227,11 +227,11 @@ function innerDiffNode(dom, vchildren, context, mountAll, absorb) {
 			else if (!child && min<childrenLen) {
 				for (j=min; j<childrenLen; j++) {
 					c = children[j];
-					if (c && isSameNodeType(c, vchild)) {
+					if (isSameNodeType(c, vchild)) {
 						child = c;
-						children[j] = undefined;
-						if (j===childrenLen-1) childrenLen--;
-						if (j===min) min++;
+						children[j] = children[min];
+						children[min] = undefined;
+						min++;
```

Comment:
To explain this and check my understanding:

1. `children` is the list of child DOM nodes which were created based on a vdom Element in the previous render but did not have a `key` prop assigned.
2. This loop iterates over the unkeyed vdom elements and unkeyed DOM elements and matches them up based on type. Once a DOM element is chosen for re-use, it can't be used for another vdom element, so the code previously set the array item to `undefined` so that it would be skipped over when processing other vdom nodes.
3. There is a cost to checking the falsyness of `undefined` entries in the array and skipping over them, which is what this PR set out to change.
4. The way it now works is by _moving_ the first unused element of the `children` array to the position of the 'used' DOM element and then clearing the first unused position. This means that all the `undefined` entries in the `children` array are at the start in positions less than `min`. Since the loop starts iterating with `j=min`, `c = children[j]` will never be undefined and the check can be removed.

A side effect is that elements in the `children` array can get reordered as the list of vdom children is processed.